src/GafferScene/Constraint.cpp

```diff
@@ -38,11 +38,594 @@
+				}
+				break;
+		}
+	}
+
+	bool valid() const
+	{
+		return ( m_uvs != 0 ) && ( m_indices != 0 );
+	}
+
+	const Imath::V2f& operator[]( const int i ) const
+	{
+		assert( valid() );
+
+		return ( *m_uvs )[ ( m_uvIndices != 0 )
+			? ( ( *m_uvIndices )[ ( *m_indices )[ i ] ] )
+			: ( ( *m_indices )[ i ] ) ];
+	}
+
+private:
+
+	const std::vector< int >* m_indices;
+	const std::vector< int >* m_uvIndices;
+	const std::vector< Imath::V2f >* m_uvs;
+};
+
+struct Index
+: boost::intrusive::list_base_hook< boost::intrusive::link_mode< boost::intrusive::normal_link > >
+{
+	int index;
+};
+
+typedef boost::intrusive::list< Index,
+	boost::intrusive::constant_time_size< true >,
+	boost::intrusive::size_type< int > > IndexList;
+
+typedef IECore::CircularIterator< IndexList::iterator, IndexList > CircularIndexIterator;
+
+void computePrimitiveVertexLocalFrame( const IECoreScene::Primitive& primitive, Imath::M44f& m, const int vertexId, const bool throwOnError )
+{
+	const IECore::V3fVectorData* const pdata = primitive.variableData< IECore::V3fVectorData >( "P" );
+	if( pdata == 0 )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( "Gaffer::Contraint : Primitive has no Vertex \"P\" primitive variable." );
+		}
+		else
+		{
+			return;
+		}
+	}
+	const IECore::V3fVectorData::ValueType& points = pdata->readable();
+
+	if( ( vertexId < 0 ) || ( vertexId >= static_cast< int >( points.size() ) ) )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( (
+				boost::format(
+					"Gaffer::Constraint : Vertex id \"%d\" is out of range."
+				) % ( vertexId ) ).str() );
+		}
+		else
+		{
+			return;
+		}
+	}
+
+	m.translate( points[ vertexId ] );
+}
+
+void computeMeshVertexLocalFrame( const IECoreScene::MeshPrimitive& primitive, Imath::M44f& m, const int vertexId, const std::string& uvSet, const bool throwOnError, const IECore::Canceller* const canceller )
+{
+	const IECore::V3fVectorData* const pdata = primitive.variableData< IECore::V3fVectorData >( "P" );
+
+	if( pdata == 0 )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( "Gaffer::Contraint : MeshPrimitive has no Vertex \"P\" primitive variable." );
+		}
+		else
+		{
+			return;
+		}
+	}
+
+	const IECore::V3fVectorData::ValueType& points = pdata->readable();
+
+	if( ( vertexId < 0 ) || ( vertexId >= static_cast< int >( points.size() ) ) )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( (
+				boost::format(
+					"Gaffer::Constraint : Vertex id \"%d\" is out of range."
+				) % ( vertexId ) ).str() );
+		}
+		else
+		{
+			return;
+		}
+	}
+
+	const IECore::IntVectorData::ValueType& indices = primitive.vertexIds()->readable();
+	const IECore::IntVectorData::ValueType& faces = primitive.verticesPerFace()->readable();
+
+	const UVIndexer uvs( primitive, uvSet, throwOnError );
+
+	Imath::V3f t( 0.f );
+	Imath::V3f b( 0.f );
+
+	if( uvs.valid() )
+	{
+		for( int i = 0, nf = faces.size(), js = 0; i < nf; ++i )
```

Comment:
I would tend to write the simpler version:
`for( int i = 0, js = 0; i < faces.size(); ++i )`

Do you think I should actually be worrying about accessing the size of a vector incurring some cost that the compiler is unable to cache unless told explicitly?

I guess in this case, `i` isn't used explicitly, so it would be possible to instead do the very succinct version:
```
int js = 0;
for( int ni : faces )
```